# <center>University of Arizona Curricula - Computer Science and Engineering Program</center>

This notebook contains a preliminary analyses of the curricula and degree plan associated with the proposed undergraduate Computer Science and Engineering program in the College of Engineering at the University of Arizona.  In addition, we provide some comparisons to other undergraduate Computer Science programs around the country using data collected as a part of a prior study.

In [10]:
using CurricularAnalytics, CurricularVisualization, CurricularOptimization

## ABET-Accredited Computer Science and Engineering Programs

The program-level learning outcomes, as specified in the ABET Criteria for Accrediting Computing Programs 2022 – 2023, are as follows:

"Graduates of the program will have an ability to:
1. Analyze a complex computing problem and to apply principles of computing andother relevant disciplines to identify solutions.
2. Design, implement, and evaluate a computing-based solution to meet a given set of computing requirements in the context of the program’s discipline.
3. Communicate effectively in a variety of professional contexts.
4. Recognize professional responsibilities and make informed judgments in
computingpractice based on legal and ethical principles.
5. Function effectively as a member or leader of a team engaged in activities appropriate to the program’s discipline."

Criterion 5 within the same ABET document stipulates the following curricular requirements:

"The program must include mathematics appropriate to the discipline and at least 30 semester credit hours (or equivalent) of up-to-date coverage of fundamental and advanced computing topics that provide both breadth and depth. The computing topics must include:
1. Techniques, skills, and tools necessary for computing practice.
2. Principles and practices for secure computing.
3. Local and global impacts of computing solutions on individuals, organizations, and society."

Additional program-specific curricular requirements, for programs with "Computer Science" as a part of their name, include:

"The curriculum requirements are in addition to the General Criteria curriculum requirements and specify topics, but do not prescribe specific courses. These requirements are:

"(a) Computer science: At least 40 semester credit hours (or equivalent) that must include:
1. Substantial coverage of algorithms and complexity, computer science theory, concepts of programming languages, and software development.
2. Substantial coverage of at least one general-purpose programming language.
3. Exposure to computer architecture and organization, information management, networking and communication, operating systems, and parallel and distributed computing.
4. The study of computing-based systems at varying levels of abstraction.
5. A major project that requires integration and application of knowledge and skills acquired in earlier course work.

(b) Mathematics: At least 15 semester credit hours (or equivalent) that must include discrete mathematics and must have mathematical rigor at least equivalent to introductory calculus. The additional mathematics might include course work in areas such as calculus, linear algebra, numerical methods, probability, statistics, or number theory.

(c) At least six semester credit hours (or equivalent) in natural science course work intended for science and engineering majors. This course work must develop an understanding of the scientific method and must include laboratory work."

The following curriculum was constructed to satisfy the aforementioned requirements.

In [11]:
CSE_curric = read_csv("./programs/college_of_engineering/Univ_of_Arizona-CSE.csv")
visualize(CSE_curric, notebook=true)

WebIO.Scope(WebIO.Node{WebIO.DOM}(WebIO.DOM(:html, :iframe), Any[""], Dict{Symbol, Any}(:src => "https://curricula-api-embed.damoursystems.com/", :id => "curriculum", :events => Dict{Symbol, WebIO.JSString}(:load => WebIO.JSString("(function (){this.contentWindow.postMessage({\"options\":{\"hideTerms\":true,\"edit\":false},\"curriculum\":{\"name\":\"Computer Science and Engineering\",\"dp_name\":\"\",\"curriculum_terms\":[{\"name\":\"Term 1\",\"curriculum_items\":[{\"curriculum_requisites\":[],\"name\":\"ENGL   101\",\"metrics\":{},\"nameSub\":\"First-Year Composition\",\"id\":2,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"UNIV 101\",\"metrics\":{},\"nameSub\":\"Intro to General Ed\",\"id\":4,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"CSE 101\",\"metrics\":{},\"nameSub\":\"Programming I\",\"id\":17,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENGR 102\",\"metrics\":{},\"nameSub\":\"Intro to Engineering\",\"id\":1,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"MATH 122A/B\",\"metrics\":{},\"nameSub\":\"Calculus I\",\"id\":13,\"credits\":5,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"UNIV 301\",\"metrics\":{},\"nameSub\":\"Gen Ed Portfolio\",\"id\":12,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ECE 311\",\"metrics\":{},\"nameSub\":\"Eng Ethics\",\"id\":26,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENGR 498A\",\"metrics\":{},\"nameSub\":\"Cross-disciplinary Design\",\"id\":28,\"credits\":3,\"nameCanonical\":\"\"}],\"id\":1},{\"name\":\"Term 2\",\"curriculum_items\":[{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{},\"nameSub\":\"GE Core: Building Connections\",\"id\":9,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":3,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ENGL   102\",\"metrics\":{},\"nameSub\":\"First-Year Composition\",\"id\":3,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":17,\"target_id\":20,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSE 102\",\"metrics\":{},\"nameSub\":\"Programming II\",\"id\":20,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":13,\"target_id\":14,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MATH 129\",\"metrics\":{},\"nameSub\":\"Calculus II\",\"id\":14,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":14,\"target_id\":15,\"type\":\"CurriculumCorequisite\"}],\"name\":\"MATH 243\",\"metrics\":{},\"nameSub\":\"Discrete Math\",\"id\":15,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":17,\"target_id\":19,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ECE 274A\",\"metrics\":{},\"nameSub\":\"Digital Logic\",\"id\":19,\"credits\":4,\"nameCanonical\":\"\"}],\"id\":2},{\"name\":\"Term 3\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":28,\"target_id\":29,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ENGR 498B\",\"metrics\":{},\"nameSub\":\"Cross-disciplinary Design\",\"id\":29,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":14,\"target_id\":16,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"SIE 305\",\"metrics\":{},\"nameSub\":\"Intro to Eng Prob/Stats \",\"id\":16,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":19,\"target_id\":21,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ECE 369A\",\"metrics\":{},\"nameSub\":\"Fundamentals of Computer Org\",\"id\":21,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{},\"nameSub\":\"GE Core: Building Connections\",\"id\":10,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{},\"nameSub\":\"GE Core: Building Connections\",\"id\":11,\"credits\":3,\"nameCanonical\":\"\"}

### Programming I Course
Notice that we are proposing a new Programming I course in this curriculum. The ability of a student to succeed in a computing-intensive major is highly dependent upon success in the first programming class. Nationally, this course often serves as a gateway to the major that tends to dissuade both women and underrepresented students from pursing a degree in a computing discipline. Thus, we propose the creation of a new introductory programming course where students will be exposed to basic concepts (e.g., programming models, problem solving, abstraction, decomposition) through the learning of a programming language, and the building of software artifacts using that language. This will be a five-credit high-impact course, designed to as to foster the equitable success of all students interested in pursuing a degree in computer science and engineering.  The additional credits in this course, beyond the typical three, will be aimed at building the student success skills, mindset, and dispositions necessary to succeed in this major. 

The existing introductory programming courses offered at the University of Arizona cannot be utilized in this program for several reasons. First, the ECE 175: Programming for Engineering Applications does not match the needs of the program because it has Calculus I as a co-requisite.  The Computer Science Curricula 2013 report created by the Association for Computer Machinery (ACM) and the IEEE Computer Society clearly indicates that Calculus I is not necessary for students to succeed in a first programming course. Thus, our sense is that adding Calculus I as a prerequisite or co-requisite would only serves to “filter” the students who are able to take this course in the first semester (when it should be take).  This, in effect, would create inequities that would only serve to reduce the diversity of the students in the program. Students in this program can learn Calculus I after learning how to program, and still graduate in four years, if we structure the curriculum appropriately. 

The CSC 110: Computer Programming I course cannot be utilized for a few reasons.  First, this course has high enrollments, and it may be difficult to accommodate additional students. More importantly, this course has one of the lowest success (i.e., pass) rates on campus. In order to create a Computer Science and Engineering program that is equity-centered, this course cannot be utilized.

Another important benefit of creating and high-efficacy Programming I class is that it can serve as an entry point to computing for student in other majors. 

The following is taken from the 2013 ACM/IEEE Curricula Study:

“An argument can be made that computer science is becoming one of the core disciplines of a 21st century university education, that is, something that any educated individual must possess some level of proficiency and understanding.”

“While CS2013 provides guidelines for undergraduate programs in computer science, we believe it is important for departments to provide computing education across a broad range of subject areas. To this end, computing departments may consider providing courses, especially at the introductory level, which are accessible and attractive to students from many disciplines. This also serves the dual purpose of attracting more students to the computing field who may not have had an initial inclination otherwise.

More broadly, as computing becomes an essential tool in other disciplines, it benefits computer science departments to be “outward facing,” building bridges to other departments and curriculum areas, encouraging students to engage in multidisciplinary work, and promoting programs that span computer science and other fields of study (for example, programs in “Computational X,” where X represents other disciplines such as biology or economics).”


### Curriculum Metrics

The `basic_metrics()` function can be used to output a set of basic metrics associated with a curriculum. Here are the basic curricular metrics associated with the University of Arizona Computer Science and Engineering program:

In [12]:
metrics = basic_metrics(CSE_curric)
println(String(take!(metrics)))


University of Arizona 
Curriculum: Computer Science and Engineering
  credit hours = 120
  number of courses = 40
  Blocking Factor --
    entire curriculum = 65
    max. value = 12, for course(s): CSE 101 - Programming I
  Centrality --
    entire curriculum = 165
    max. value = 40, for course(s): Algorithm Design & Analysis
  Delay Factor --
    entire curriculum = 114.0
    max. value = 5.0, for course(s): MATH 122A/B - Calculus I, MATH 129 - Calculus II, MATH 243 - Discrete Math, CSE 101 - Programming I, SFWE 401 - Software Assurance/Security, CSE 102 - Programming II, Intro to Data Structures, Algorithm Design & Analysis, SFWE 302 - Software Design, Upper Division Computing Elective, Upper Division Computing Elective, Upper Division Computing Elective, General Elective
  Complexity --
    entire curriculum = 179.0
    max. value = 17.0, for course(s): CSE 101 - Programming I
  Longest Path(s) --
    length = 5, number of paths = 9
    path(s):
    path 1 = CSE 101 - Programming

Note that this is a 120 credit hour ABET-satsifying curriculum. From a student progression perspective, Math 122A/B appears to be the most important course in the curriculum. All long pathways in this curriculum are the math pathway. 

### Extraneous Prerequisites
The following function will find prerequisites in a curriculum.  These are redundant prerequisites that are unnecessary in a curriculum.  For example, if a curriculum has the prerequisite 
relationships $c_1 \rightarrow c_2 \rightarrow c_3$ and $c_1 \rightarrow c_3$, and $c_1$ and $c_2$ are 
*not* co-requisites, then $c_1 \rightarrow c_3$ is redundant and therefore extraneous.  Extraneous prerequisites do not effect the curricular complexity metric, they simply are unnecessary clutter in a curriculum or degree plan.

In [20]:
extraneous_requisites(CSE_curric, print=true)

Vector{Int64}[]

There are no extraneous prerequisites in this Computer Science and Engineering curriculum.

### Dead End Courses
The following function can be used to find "dead end" courses in a curricula.  Dead end courses are those that appear at the end of a path (i.e., sink vertices), and are not a part of a course associated with the major.  E.g., in the case of the CS curriculum above, these would be courses at the end of a path that do not have the "CSE" or "ENGR" prefixes.  One might consider these courses dead ends, as their course outcomes are not (formally) used by any major-specific course, i.e., by any course with the prefix "CSE."

In [21]:
de = dead_ends(CSE_curric, ["CSE", "ENGR"])
println("\nDead end courses in the $(CSE_curric.name) curriculum:")
for course in de[2]
    println("$(course.prefix) $(course.num): $(course.name)")
end


Dead end courses in the Computer Science and Engineering curriculum:
ENGL   102: First-Year Composition
SIE 305: Intro to Eng Prob/Stats 
SFWE 401: Software Assurance/Security


English Composition II is a general education course, and the skills developed in this class are informally used in other classes in the curriculum, particularly those that are writing intensive. The Software Assurance/Security course is from a closely realated program (that will be in the same department). This is not an issue.

### Comparison to CS Program in College of Science
The current curriclum/degree plan offered by the Computer Science program in the College of Science is provided below.

In [15]:
CS_plan = read_csv("./programs/college_of_science/Univ_of_Arizona-CS.csv")
visualize(CS_plan, notebook=true)

WebIO.Scope(WebIO.Node{WebIO.DOM}(WebIO.DOM(:html, :iframe), Any[""], Dict{Symbol, Any}(:src => "https://curricula-api-embed.damoursystems.com/", :id => "curriculum", :events => Dict{Symbol, WebIO.JSString}(:load => WebIO.JSString("(function (){this.contentWindow.postMessage({\"options\":{\"hideTerms\":false,\"edit\":false},\"curriculum\":{\"name\":\"Computer Science\",\"dp_name\":\"2019-20 Degree Plan\",\"curriculum_terms\":[{\"name\":\"Term 1\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":1,\"type\":\"CurriculumCorequisite\"}],\"name\":\"CSC 110\",\"metrics\":{\"centrality\":25,\"complexity\":12.0,\"blocking factor\":7,\"delay factor\":5.0},\"nameSub\":\"Intro to Computer Programming I\",\"id\":1,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"MATH 120R\",\"metrics\":{\"centrality\":0,\"complexity\":15.0,\"blocking factor\":10,\"delay factor\":5.0},\"nameSub\":\"Pre-Calculus\",\"id\":2,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENGL 101\",\"metrics\":{\"centrality\":0,\"complexity\":3.0,\"blocking factor\":1,\"delay factor\":2.0},\"nameSub\":\"English Composition I\",\"id\":3,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":3.0,\"blocking factor\":1,\"delay factor\":2.0},\"nameSub\":\"1st Semester Second Language\",\"id\":4,\"credits\":4,\"nameCanonical\":\"\"}],\"id\":1},{\"name\":\"Term 2\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":1,\"target_id\":5,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSC 120\",\"metrics\":{\"centrality\":25,\"complexity\":11.0,\"blocking factor\":6,\"delay factor\":5.0},\"nameSub\":\"Intro to Computer Programming II\",\"id\":5,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":6,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MATH 122A/B\",\"metrics\":{\"centrality\":3,\"complexity\":4.0,\"blocking factor\":1,\"delay factor\":3.0},\"nameSub\":\"Calculus I\",\"id\":6,\"credits\":5,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":3,\"target_id\":7,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ENGL 102\",\"metrics\":{\"centrality\":0,\"complexity\":2.0,\"blocking factor\":0,\"delay factor\":2.0},\"nameSub\":\"English Composition II\",\"id\":7,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":4,\"target_id\":8,\"type\":\"CurriculumPrerequisite\"}],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":2.0,\"blocking factor\":0,\"delay factor\":2.0},\"nameSub\":\"2nd Semester Second Language\",\"id\":8,\"credits\":4,\"nameCanonical\":\"\"}],\"id\":2},{\"name\":\"Term 3\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":5,\"target_id\":9,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSC 210\",\"metrics\":{\"centrality\":20,\"complexity\":9.0,\"blocking factor\":4,\"delay factor\":5.0},\"nameSub\":\"Software Development\",\"id\":9,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":5,\"target_id\":10,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSC 245\",\"metrics\":{\"centrality\":5,\"complexity\":6.0,\"blocking factor\":1,\"delay factor\":5.0},\"nameSub\":\"Intro to Discrete Structures\",\"id\":10,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":6,\"target_id\":11,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MATH 129\",\"metrics\":{\"centrality\":0,\"complexity\":3.0,\"blocking factor\":0,\"delay factor\":3.0},\"nameSub\":\"Calculus II\",\"id\":11,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Tier 1 General Education 1\",\"id\":12,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"bloc

In [16]:
metrics = basic_metrics(CS_plan.curriculum)
println(String(take!(metrics)))


University of Arizona 
Curriculum: Computer Science
  credit hours = 122
  number of courses = 37
  Blocking Factor --
    entire curriculum = 31
    max. value = 10, for course(s): MATH 120R - Pre-Calculus
  Centrality --
    entire curriculum = 78
    max. value = 25, for course(s): CSC 110 - Intro to Computer Programming I, CSC 120 - Intro to Computer Programming II
  Delay Factor --
    entire curriculum = 81.0
    max. value = 5.0, for course(s): CSC 110 - Intro to Computer Programming I, MATH 120R - Pre-Calculus, CSC 120 - Intro to Computer Programming II, CSC 210 - Software Development, CSC 245 - Intro to Discrete Structures, CSC 252 - Computer Organization, CSC 352 - Systems Programming & Unix, CSC 335 - Object-Oriented Programming, CSC 345 - Analysis of Discrete Structures
  Complexity --
    entire curriculum = 112.0
    max. value = 15.0, for course(s): MATH 120R - Pre-Calculus
  Longest Path(s) --
    length = 5, number of paths = 5
    path(s):
    path 1 = MATH 120R - Pr

In [17]:
metrics = basic_metrics(CS_plan)
println(String(take!(metrics)))


Curriculum: Computer Science
Degree Plan: 2019-20 Degree Plan
  total credit hours = 122
  number of terms = 8
  max. credits in a term = 17, in term 2
  min. credits in a term = 12, in term 8
  avg. credits per term = 15.25, with std. dev. = 1.3919410907075054



The differences between the introductory programming courses in these two curricula was already considered above. 

The CS program above contains nine credit hours of mathemtics beginning with Calculus I, and include discrete mathematics. The engineering-based CSE program includes fifteen credit hours of mathematics beginning with Calculus I, including discrete mathematics and probability theory, as well as one three-credit-hour mathematics class that builds on Calculus II, which may include linear algebra, number theory, numerical methods.  These additional mathematics courses will prepare students to purse applied work related to security and cryptography, machine learning, and analysis/data science, as well as an advanced engineering degree in numerous fields.

Finally, both the CS and CSE curricula contain 42 credit hours of Computer Science topics. 

## Computer Science and Engineering Degree Plan
A possible degree plan for the Computer Science and Engineering program, subject to the various constraints provided below, is as follows:

In [19]:
fixed = Dict{Course,Int}()
Eng101 = findfirst(x->x.id==1, CSE_curric.courses)
Programming1 = findfirst(x->x.id==17, CSE_curric.courses)
UNIV101 = findfirst(x->x.id==4, CSE_curric.courses)
SeniorDesign1 = findfirst(x->x.id==28, CSE_curric.courses)
SeniorDesign2 = findfirst(x->x.id==29, CSE_curric.courses)
Comp1 = findfirst(x->x.id==2, CSE_curric.courses)
Calc1 = findfirst(x->x.id==13, CSE_curric.courses)
fixed[CSE_curric.courses[Eng101]] = 1
fixed[CSE_curric.courses[Programming1]] = 1
fixed[CSE_curric.courses[UNIV101]] = 1
fixed[CSE_curric.courses[Comp1]] = 1
fixed[CSE_curric.courses[Calc1]] = 1
fixed[CSE_curric.courses[SeniorDesign1]] = 7
fixed[CSE_curric.courses[SeniorDesign2]] = 8
CSE_plan = optimize_plan(CSE_curric, 8, 12, 18, [balance_obj, req_distance_obj], fixed_terms=fixed)
visualize(CSE_plan, notebook=true)

An optimal solution was found with objective value = 28.0


WebIO.Scope(WebIO.Node{WebIO.DOM}(WebIO.DOM(:html, :iframe), Any[""], Dict{Symbol, Any}(:src => "https://curricula-api-embed.damoursystems.com/", :id => "curriculum", :events => Dict{Symbol, WebIO.JSString}(:load => WebIO.JSString("(function (){this.contentWindow.postMessage({\"options\":{\"hideTerms\":false,\"edit\":false},\"curriculum\":{\"name\":\"Computer Science and Engineering\",\"dp_name\":\"\",\"curriculum_terms\":[{\"name\":\"Term 1\",\"curriculum_items\":[{\"curriculum_requisites\":[],\"name\":\"CSE 101\",\"metrics\":{\"centrality\":0,\"complexity\":17.0,\"blocking factor\":12,\"delay factor\":5.0},\"nameSub\":\"Programming I\",\"id\":17,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENGR 102\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Intro to Engineering\",\"id\":1,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"UNIV 101\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Intro to General Ed\",\"id\":4,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"MATH 122A/B\",\"metrics\":{\"centrality\":0,\"complexity\":15.0,\"blocking factor\":10,\"delay factor\":5.0},\"nameSub\":\"Calculus I\",\"id\":13,\"credits\":5,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENGL   101\",\"metrics\":{\"centrality\":0,\"complexity\":3.0,\"blocking factor\":1,\"delay factor\":2.0},\"nameSub\":\"First-Year Composition\",\"id\":2,\"credits\":3,\"nameCanonical\":\"\"}],\"id\":1},{\"name\":\"Term 2\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":17,\"target_id\":20,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSE 102\",\"metrics\":{\"centrality\":29,\"complexity\":14.0,\"blocking factor\":9,\"delay factor\":5.0},\"nameSub\":\"Programming II\",\"id\":20,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"UNIV 301\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Gen Ed Portfolio\",\"id\":12,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":17,\"target_id\":19,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ECE 274A\",\"metrics\":{\"centrality\":4,\"complexity\":6.0,\"blocking factor\":2,\"delay factor\":4.0},\"nameSub\":\"Digital Logic\",\"id\":19,\"credits\":4,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":3,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ENGL   102\",\"metrics\":{\"centrality\":0,\"complexity\":2.0,\"blocking factor\":0,\"delay factor\":2.0},\"nameSub\":\"First-Year Composition\",\"id\":3,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":3.0,\"blocking factor\":1,\"delay factor\":2.0},\"nameSub\":\"GE Core: Exploring Perspectives (Natural Scientist)\",\"id\":7,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ECE 311\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Eng Ethics\",\"id\":26,\"credits\":1,\"nameCanonical\":\"\"}],\"id\":2},{\"name\":\"Term 3\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":20,\"target_id\":22,\"type\":\"CurriculumPrerequisite\"}],\"name\":\" \",\"metrics\":{\"centrality\":29,\"complexity\":13.0,\"blocking factor\":8,\"delay factor\":5.0},\"nameSub\":\"Intro to Data Structures\",\"id\":22,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":13,\"target_id\":14,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MATH 129\",\"metrics\":{\"centrality\":30,\"complexity\":14.0,\"blocking factor\":9,\"delay factor\":5.0},\"nameSub\":\"Calculus II\",\"id\":14,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":19,\"target_id\":21,\"type\":\"CurriculumPrerequisite\"}

Note that it is easily possible to create a four-year degree plan with Calculus I starting in the second term.  That is, this program easily accommodates students who do *not* test into Calculus I as their math starting point.

### Degree Plan Metrics

The metrics for this degree plan are as follows:

In [56]:
metrics = basic_metrics(CSE_plan)
println(String(take!(metrics)))


Curriculum: Computer Science and Engineering
Degree Plan: 
  total credit hours = 120
  number of terms = 8
  max. credits in a term = 16, in term 1
  min. credits in a term = 14, in term 3
  avg. credits per term = 15.0, with std. dev. = 0.5

